### PR TITLE
Coffee

### LINKED ISSUE #295: Close the gap between recruit and make_arrangements.
Currently, Alice is able to `.recruit()`, using `sample()`, `n` Ursulas, weighed by stake, and indicated by ethereum address.

However, there is no authoritative way to get the REST details from this information, which Alice needs to move on to `make_arrangements()`.

### PR DESCRIPTION
**Pull request by jMyles**
#### The initial goals for this PR were as follows:

- The highest-level goal of this PR is to resolve #295.  Doing that also requires:
-- Ursula becoming "fully available" for Alice to find, as described in #294.
-- Complete adjustment of Arrangement ID as described in #180 
-- Versioned Arrangements #281 
-- Better node caching on Alice, perhaps sharing logic with #227  

We have pushed #180 and #281 down the road a bit, but the rest are closed, along with many others along the way (see the various included PRs for details: #307, #308, #318, #324, #327, #330, #336, #344, #345, #348, and #351).

#### What this PR represents
Closing the gap between Alice having access to the blockchain and actually being able to create real Policies - the core of the NuCypher offering - is one of the crucial details to work out to realize this project's goals.  This PR specifically addresses the portions of that gap involving discovering, remembering, and verifying nodes (especially Ursulas).

As you read this PR and consider it for merge, consider these as the main points of change:

#### Character
- All actors now learn about nodes now learn about each other in a uniform fashion.
- Learning happens off the main thread in a "learning loop."
- All Characters now have a canonical address.
- DHT legality logic has blossomed into a full-blown verification scheme for Ursulas, and is used both by other Ursulas when deciding whether to propagate and also by other actors when deciding when to interact with Ursula.

#### Policy
- Creates Arrangements and Policies from sampled addresses
- Deprecate mocked blockchain client
- Create Arrangements from sampled ursulas
- Enact and publish policies via rest and blockchain network
- Invert `Arrangement` and `BlockchainArrangement` inheritance order

#### Network
- Deprecates `NuCypherDHTNode` and `Capabilities`
- Renames `NetworkyStuff` -> `NetworkMiddleware`
- Eliminates DHT usage entirely for Alice and Bob (although Alice still has some entirely optional DHT capability)

#### Blockchain
- New stake logic; stake dividing is possible, unlocking is stepwise.
- Embed default web3.py providers in Interface
- Use Interface-style API for Blockchain connection
- Reduces complexity of `ContractAgent` API with default values
- Implements miner's datastore publish and fetch agent methods
- Uses Ursula's ethereum keypair to sign and verify her other metadata
- Improves solidity compile-time exception handling
- Optimzes contract source code
- Introduce blockchain actors to existing tests, fixtures, and utilities

#### Configuration
- Compose `NucypherConfiguration` and `BlockchainConfiguration`
- Decouple configuration: Keep it optional
- Refactors `NucypherMinerConstants` into a module that uses constant sorrow + side effects

### CODE REVIEW

File: examples/finnegans-wake-federated.py [hunk 1]
```diff
@@ -133,9 +138,10 @@
 
     # Now Bob can retrieve the original message.  He just needs the MessageKit
     # and the DataSource which produced it.
+    alice_pubkey_restored_from_ancient_scroll = UmbralPublicKey.from_bytes(alices_pubkey_bytes_saved_for_posterity)
     delivered_cleartext = BOB.retrieve(message_kit=message_kit,
                                        data_source=datasource_as_understood_by_bob,
-                                       alice_pubkey_sig=alices_pubkey_saved_for_posterity)
+                                       alice_pubkey_sig=alice_pubkey_restored_from_ancient_scroll)
```
Comment (michwill) [inline examples/finnegans-wake-federated.py]: nice

File: nucypher/blockchain/eth/actors.py [hunk 2]
```diff
@@ -1,250 +1,310 @@
...
     def is_staking(self):
         """Checks if this Miner currently has locked tokens."""
-
-        self.__update_locked_tokens()
-        return bool(self.__locked_tokens > 0)
+        return bool(self.locked_tokens > 0)
 
     @property
-    def locked_tokens(self,):
+    def locked_tokens(self, ):
```
Comment (jMyles) [inline nucypher/blockchain/eth/actors.py]: What ahhh... what are we looking at here?

File: nucypher/blockchain/eth/actors.py [hunk 3]
```diff
@@ -1,250 +1,310 @@
...
+    def stakes(self):
+        stakes_reader = self.miner_agent.get_all_stakes(miner_address=self.checksum_public_address)
+        return stakes_reader
 
+    @only_me
     def deposit(self, amount: int, lock_periods: int) -> Tuple[str, str]:
         """Public facing method for token locking."""
-        approve_txhash = self._approve_escrow(amount=amount)
-        deposit_txhash = self._send_tokens_to_escrow(amount=amount, lock_periods=lock_periods)
+        if not self.is_me:
```
Comment (jMyles) [inline nucypher/blockchain/eth/actors.py]: This is unreachable now because of the decorator.

File: nucypher/blockchain/eth/actors.py (hunk 2 above)
Comment (michwill) [reply nucypher/blockchain/eth/actors.py]: hah

File: nucypher/blockchain/eth/agents.py [hunk 4]
```diff
@@ -78,105 +81,200 @@ class MinerAgent(EthereumContractAgent, NucypherMinerConstants):
...
+            miner_address = self.contract.functions.miners(index).call()
+            yield miner_address
 
-    def sample(self, quantity: int=10, additional_ursulas: float=1.7, attempts: int=5, duration: int=10) -> List[str]:
+    def sample(self, quantity: int, duration: int, additional_ursulas: float=1.7, attempts: int=5) -> List[str]:
         """
         Select n random staking Ursulas, according to their stake distribution.
         The returned addresses are shuffled, so one can request more than needed and
         throw away those which do not respond.
-
```
Comment (KPrasch) [inline nucypher/blockchain/eth/agents.py]: Did we loose this chart?

**Pull request merged**